// @effect/cli/dist/esm/BuiltInOptions.js
/**
 * @since 1.0.0
 */
import * as InternalBuiltInOptions from "./internal/builtInOptions.js";
/**
 * @since 1.0.0
 * @category options
 */
export const builtInOptions = InternalBuiltInOptions.builtInOptions;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isShowCompletions = InternalBuiltInOptions.isShowCompletions;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isShowHelp = InternalBuiltInOptions.isShowHelp;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isShowWizard = InternalBuiltInOptions.isShowWizard;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isShowVersion = InternalBuiltInOptions.isShowVersion;
/**
 * @since 1.0.0
 * @category constructors
 */
export const showCompletions = InternalBuiltInOptions.showCompletions;
/**
 * @since 1.0.0
 * @category constructors
 */
export const showHelp = InternalBuiltInOptions.showHelp;
/**
 * @since 1.0.0
 * @category constructors
 */
export const showWizard = InternalBuiltInOptions.showWizard;
/**
 * @since 1.0.0
 * @category constructors
 */
export const showVersion = InternalBuiltInOptions.showVersion;
