// @effect/cli/dist/esm/CommandDescriptor.js
import * as Internal from "./internal/commandDescriptor.js";
/**
 * @since 1.0.0
 * @category symbols
 */
export const TypeId = Internal.TypeId;
/**
 * @since 1.0.0
 * @category combinators
 */
export const getHelp = Internal.getHelp;
/**
 * @since 1.0.0
 * @category combinators
 */
export const getBashCompletions = Internal.getBashCompletions;
/**
 * @since 1.0.0
 * @category combinators
 */
export const getFishCompletions = Internal.getFishCompletions;
/**
 * @since 1.0.0
 * @category combinators
 */
export const getZshCompletions = Internal.getZshCompletions;
/**
 * @since 1.0.0
 * @category combinators
 */
export const getNames = Internal.getNames;
/**
 * @since 1.0.0
 * @category combinators
 */
export const getSubcommands = Internal.getSubcommands;
/**
 * @since 1.0.0
 * @category combinators
 */
export const getUsage = Internal.getUsage;
/**
 * @since 1.0.0
 * @category combinators
 */
export const map = Internal.map;
/**
 * @since 1.0.0
 * @category combinators
 */
export const mapEffect = Internal.mapEffect;
/**
 * @since 1.0.0
 * @category combinators
 */
export const parse = Internal.parse;
/**
 * @since 1.0.0
 * @category constructors
 */
export const prompt = Internal.prompt;
/**
 * @since 1.0.0
 * @category constructors
 */
export const make = Internal.make;
/**
 * @since 1.0.0
 * @category combinators
 */
export const withDescription = Internal.withDescription;
/**
 * @since 1.0.0
 * @category combinators
 */
export const withSubcommands = Internal.withSubcommands;
/**
 * @since 1.0.0
 * @category combinators
 */
export const wizard = Internal.wizard;
