// @effect/cli/dist/esm/Command.js
import * as Internal from "./internal/command.js";
/**
 * @since 1.0.0
 * @category type ids
 */
export const TypeId = Internal.TypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
export const fromDescriptor = Internal.fromDescriptor;
/**
 * @since 1.0.0
 * @category accessors
 */
export const getHelp = Internal.getHelp;
/**
 * @since 1.0.0
 * @category accessors
 */
export const getNames = Internal.getNames;
/**
 * @since 1.0.0
 * @category accessors
 */
export const getBashCompletions = Internal.getBashCompletions;
/**
 * @since 1.0.0
 * @category accessors
 */
export const getFishCompletions = Internal.getFishCompletions;
/**
 * @since 1.0.0
 * @category accessors
 */
export const getZshCompletions = Internal.getZshCompletions;
/**
 * @since 1.0.0
 * @category accessors
 */
export const getSubcommands = Internal.getSubcommands;
/**
 * @since 1.0.0
 * @category accessors
 */
export const getUsage = Internal.getUsage;
/**
 * @since 1.0.0
 * @category constructors
 */
export const make = Internal.make;
/**
 * @since 1.0.0
 * @category constructors
 */
export const prompt = Internal.prompt;
/**
 * @since 1.0.0
 * @category combinators
 */
export const provide = Internal.provide;
/**
 * @since 1.0.0
 * @category combinators
 */
export const provideEffect = Internal.provideEffect;
/**
 * @since 1.0.0
 * @category combinators
 */
export const provideEffectDiscard = Internal.provideEffectDiscard;
/**
 * @since 1.0.0
 * @category combinators
 */
export const provideSync = Internal.provideSync;
/**
 * @since 1.0.0
 * @category combinators
 */
export const transformHandler = Internal.transformHandler;
/**
 * @since 1.0.0
 * @category combinators
 */
export const withDescription = Internal.withDescription;
/**
 * @since 1.0.0
 * @category combinators
 */
export const withHandler = Internal.withHandler;
/**
 * @since 1.0.0
 * @category combinators
 */
export const withSubcommands = Internal.withSubcommands;
/**
 * @since 1.0.0
 * @category accessors
 */
export const wizard = Internal.wizard;
/**
 * @since 1.0.0
 * @category conversions
 */
export const run = Internal.run;
