// @effect/cli/dist/esm/HelpDoc.js
import * as InternalHelpDoc from "./internal/helpDoc.js";
/**
 * @since 1.0.0
 * @category refinements
 */
export const isEmpty = InternalHelpDoc.isEmpty;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isHeader = InternalHelpDoc.isHeader;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isParagraph = InternalHelpDoc.isParagraph;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isDescriptionList = InternalHelpDoc.isDescriptionList;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isEnumeration = InternalHelpDoc.isEnumeration;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isSequence = InternalHelpDoc.isSequence;
/**
 * @since 1.0.0
 * @category constructors
 */
export const empty = InternalHelpDoc.empty;
/**
 * @since 1.0.0
 * @category constructors
 */
export const blocks = InternalHelpDoc.blocks;
/**
 * @since 1.0.0
 * @category constructors
 */
export const h1 = InternalHelpDoc.h1;
/**
 * @since 1.0.0
 * @category constructors
 */
export const h2 = InternalHelpDoc.h2;
/**
 * @since 1.0.0
 * @category constructors
 */
export const h3 = InternalHelpDoc.h3;
/**
 * @since 1.0.0
 * @category constructors
 */
export const p = InternalHelpDoc.p;
/**
 * @since 1.0.0
 * @category constructors
 */
export const descriptionList = InternalHelpDoc.descriptionList;
/**
 * @since 1.0.0
 * @category constructors
 */
export const enumeration = InternalHelpDoc.enumeration;
/**
 * @since 1.0.0
 * @category getters
 */
export const getSpan = InternalHelpDoc.getSpan;
/**
 * @since 1.0.0
 * @category combinators
 */
export const sequence = InternalHelpDoc.sequence;
/**
 * @since 1.0.0
 * @category combinators
 */
export const orElse = InternalHelpDoc.orElse;
/**
 * @since 1.0.0
 * @category mapping
 */
export const mapDescriptionList = InternalHelpDoc.mapDescriptionList;
/**
 * @since 1.0.0
 * @category rendering
 */
export const toAnsiDoc = InternalHelpDoc.toAnsiDoc;
/**
 * @since 1.0.0
 * @category rendering
 */
export const toAnsiText = InternalHelpDoc.toAnsiText;
