// @effect/platform/dist/esm/OpenApi.js
/**
 * @since 1.0.0
 */
import * as Context from "effect/Context";
import { constFalse } from "effect/Function";
import { globalValue } from "effect/GlobalValue";
import * as Option from "effect/Option";
import * as HttpApi from "./HttpApi.js";
import * as HttpApiMiddleware from "./HttpApiMiddleware.js";
import * as HttpApiSchema from "./HttpApiSchema.js";
import * as HttpMethod from "./HttpMethod.js";
import * as JsonSchema from "./OpenApiJsonSchema.js";
/**
 * @since 1.0.0
 * @category annotations
 */
export class Identifier extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Identifier")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Title extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Title")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Version extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Version")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Description extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Description")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class License extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/License")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class ExternalDocs extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/ExternalDocs")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Servers extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Servers")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Format extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Format")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Summary extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Summary")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Deprecated extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Deprecated")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Override extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Override")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
export class Exclude extends /*#__PURE__*/Context.Reference()("@effect/platform/OpenApi/Exclude", {
  defaultValue: constFalse
}) {}
/**
 * Transforms the generated OpenAPI specification
 * @since 1.0.0
 * @category annotations
 */
export class Transform extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Transform")() {}
const contextPartial = tags => {
  const entries = Object.entries(tags);
  return options => {
    let context = Context.empty();
    for (const [key, tag] of entries) {
      if (options[key] !== undefined) {
        context = Context.add(context, tag, options[key]);
      }
    }
    return context;
  };
};
/**
 * @since 1.0.0
 * @category annotations
 */
export const annotations = /*#__PURE__*/contextPartial({
  identifier: Identifier,
  title: Title,
  version: Version,
  description: Description,
  license: License,
  summary: Summary,
  deprecated: Deprecated,
  externalDocs: ExternalDocs,
  servers: Servers,
  format: Format,
  override: Override,
  exclude: Exclude,
  transform: Transform
});
const apiCache = /*#__PURE__*/globalValue("@effect/platform/OpenApi/apiCache", () => new WeakMap());
/**
 * @category constructors
 * @since 1.0.0
 */
export const fromApi = self => {
  if (apiCache.has(self)) {
    return apiCache.get(self);
  }
  const api = self;
  const jsonSchemaDefs = {};
  let spec = {
    openapi: "3.1.0",
    info: {
      title: Context.getOrElse(api.annotations, Title, () => "Api"),
      version: Context.getOrElse(api.annotations, Version, () => "0.0.1")
    },
    paths: {},
    components: {
      schemas: jsonSchemaDefs,
      securitySchemes: {}
    },
    security: [],
    tags: []
  };
  function makeJsonSchemaOrRef(ast) {
    return JsonSchema.fromAST(ast, {
      defs: jsonSchemaDefs
    });
  }
  function registerSecurity(name, security) {
    if (spec.components.securitySchemes[name]) {
      return;
    }
    const scheme = makeSecurityScheme(security);
    spec.components.securitySchemes[name] = scheme;
  }
  Option.map(Context.getOption(api.annotations, HttpApi.AdditionalSchemas), componentSchemas => {
    componentSchemas.forEach(componentSchema => makeJsonSchemaOrRef(componentSchema.ast));
  });
  Option.map(Context.getOption(api.annotations, Description), description => {
    spec.info.description = description;
  });
  Option.map(Context.getOption(api.annotations, License), license => {
    spec.info.license = license;
  });
  Option.map(Context.getOption(api.annotations, Summary), summary => {
    spec.info.summary = summary;
  });
  Option.map(Context.getOption(api.annotations, Servers), servers => {
    spec.servers = servers;
  });
  Option.map(Context.getOption(api.annotations, Override), override => {
    Object.assign(spec, override);
  });
  api.middlewares.forEach(middleware => {
    if (!HttpApiMiddleware.isSecurity(middleware)) {
      return;
    }
    for (const [name, security] of Object.entries(middleware.security)) {
      registerSecurity(name, security);
      spec.security.push({
        [name]: []
      });
    }
  });
  HttpApi.reflect(api, {
    onGroup({
      group
    }) {
      if (Context.get(group.annotations, Exclude)) {
        return;
      }
      let tag = {
        name: Context.getOrElse(group.annotations, Title, () => group.identifier)
      };
      Option.map(Context.getOption(group.annotations, Description), description => {
        tag.description = description;
      });
      Option.map(Context.getOption(group.annotations, ExternalDocs), externalDocs => {
        tag.externalDocs = externalDocs;
      });
      Option.map(Context.getOption(group.annotations, Override), override => {
        Object.assign(tag, override);
      });
      Option.map(Context.getOption(group.annotations, Transform), fn => {
        tag = fn(tag);
      });
      spec.tags.push(tag);
    },
    onEndpoint({
      endpoint,
      errors,
      group,
      mergedAnnotations,
      middleware,
      payloads,
      successes
    }) {
      if (Context.get(mergedAnnotations, Exclude)) {
        return;
      }
      let op = {
        tags: [Context.getOrElse(group.annotations, Title, () => group.identifier)],
        operationId: Context.getOrElse(endpoint.annotations, Identifier, () => group.topLevel ? endpoint.name : `${group.identifier}.${endpoint.name}`),
        parameters: [],
        security: [],
        responses: {}
      };
      function processResponseMap(map, defaultDescription) {
        for (const [status, {
          ast,
          description
        }] of map) {
          if (op.responses[status]) continue;
          op.responses[status] = {
            description: Option.getOrElse(description, defaultDescription)
          };
          ast.pipe(Option.filter(ast => !HttpApiSchema.getEmptyDecodeable(ast)), Option.map(ast => {
            const encoding = HttpApiSchema.getEncoding(ast);
            op.responses[status].content = {
              [encoding.contentType]: {
                schema: makeJsonSchemaOrRef(ast)
              }
            };
          }));
        }
      }
      function processParameters(schema, i) {
        if (Option.isSome(schema)) {
          const jsonSchema = makeJsonSchemaOrRef(schema.value.ast);
          if ("properties" in jsonSchema) {
            Object.entries(jsonSchema.properties).forEach(([name, psJsonSchema]) => {
              op.parameters.push({
                name,
                in: i,
                schema: psJsonSchema,
                required: jsonSchema.required.includes(name),
                ...(psJsonSchema.description !== undefined ? {
                  description: psJsonSchema.description
                } : undefined)
              });
            });
          }
        }
      }
      Option.map(Context.getOption(endpoint.annotations, Description), description => {
        op.description = description;
      });
      Option.map(Context.getOption(endpoint.annotations, Summary), summary => {
        op.summary = summary;
      });
      Option.map(Context.getOption(endpoint.annotations, Deprecated), deprecated => {
        op.deprecated = deprecated;
      });
      Option.map(Context.getOption(endpoint.annotations, ExternalDocs), externalDocs => {
        op.externalDocs = externalDocs;
      });
      middleware.forEach(middleware => {
        if (!HttpApiMiddleware.isSecurity(middleware)) {
          return;
        }
        for (const [name, security] of Object.entries(middleware.security)) {
          registerSecurity(name, security);
          op.security.push({
            [name]: []
          });
        }
      });
      const hasBody = HttpMethod.hasBody(endpoint.method);
      if (hasBody && payloads.size > 0) {
        const content = {};
        payloads.forEach(({
          ast
        }, contentType) => {
          content[contentType] = {
            schema: makeJsonSchemaOrRef(ast)
          };
        });
        op.requestBody = {
          content,
          required: true
        };
      }
      processParameters(endpoint.pathSchema, "path");
      if (!hasBody) {
        processParameters(endpoint.payloadSchema, "query");
      }
      processParameters(endpoint.headersSchema, "header");
      processParameters(endpoint.urlParamsSchema, "query");
      processResponseMap(successes, () => "Success");
      processResponseMap(errors, () => "Error");
      const path = endpoint.path.replace(/:(\w+)[^/]*/g, "{$1}");
      const method = endpoint.method.toLowerCase();
      if (!spec.paths[path]) {
        spec.paths[path] = {};
      }
      Option.map(Context.getOption(endpoint.annotations, Override), override => {
        Object.assign(op, override);
      });
      Option.map(Context.getOption(endpoint.annotations, Transform), transformFn => {
        op = transformFn(op);
      });
      spec.paths[path][method] = op;
    }
  });
  Option.map(Context.getOption(api.annotations, Transform), transformFn => {
    spec = transformFn(spec);
  });
  apiCache.set(self, spec);
  return spec;
};
const makeSecurityScheme = security => {
  const meta = {};
  Option.map(Context.getOption(security.annotations, Description), description => {
    meta.description = description;
  });
  switch (security._tag) {
    case "Basic":
      {
        return {
          ...meta,
          type: "http",
          scheme: "basic"
        };
      }
    case "Bearer":
      {
        const format = Context.getOption(security.annotations, Format).pipe(Option.map(format => ({
          bearerFormat: format
        })), Option.getOrUndefined);
        return {
          ...meta,
          type: "http",
          scheme: "bearer",
          ...format
        };
      }
    case "ApiKey":
      {
        return {
          ...meta,
          type: "apiKey",
          name: security.key,
          in: security.in
        };
      }
  }
};
