// @effect/printer-ansi/dist/cjs/internal/color.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.yellow = exports.white = exports.toCode = exports.red = exports.magenta = exports.green = exports.cyan = exports.blue = exports.black = void 0;
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/** @internal */
const black = exports.black = {
  _tag: "Black"
};
/** @internal */
const red = exports.red = {
  _tag: "Red"
};
/** @internal */
const green = exports.green = {
  _tag: "Green"
};
/** @internal */
const yellow = exports.yellow = {
  _tag: "Yellow"
};
/** @internal */
const blue = exports.blue = {
  _tag: "Blue"
};
/** @internal */
const magenta = exports.magenta = {
  _tag: "Magenta"
};
/** @internal */
const cyan = exports.cyan = {
  _tag: "Cyan"
};
/** @internal */
const white = exports.white = {
  _tag: "White"
};
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/** @internal */
const toCode = color => {
  switch (color._tag) {
    case "Black":
      {
        return 0;
      }
    case "Red":
      {
        return 1;
      }
    case "Green":
      {
        return 2;
      }
    case "Yellow":
      {
        return 3;
      }
    case "Blue":
      {
        return 4;
      }
    case "Magenta":
      {
        return 5;
      }
    case "Cyan":
      {
        return 6;
      }
    case "White":
      {
        return 7;
      }
  }
};
exports.toCode = toCode;
