// @effect/printer-ansi/dist/cjs/internal/sgr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toEscapeSequence = exports.toCode = exports.setUnderlined = exports.setStrikethrough = exports.setItalicized = exports.setColor = exports.setBold = exports.reset = void 0;
var color = _interopRequireWildcard(require("./color.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/** @internal */
const reset = exports.reset = {
  _tag: "Reset"
};
/** @internal */
const setBold = bold => ({
  _tag: "SetBold",
  bold
});
/** @internal */
exports.setBold = setBold;
const setColor = (color, vivid, layer) => ({
  _tag: "SetColor",
  color,
  vivid,
  layer
});
/** @internal */
exports.setColor = setColor;
const setItalicized = italicized => ({
  _tag: "SetItalicized",
  italicized
});
/** @internal */
exports.setItalicized = setItalicized;
const setStrikethrough = strikethrough => ({
  _tag: "SetStrikethrough",
  strikethrough
});
/** @internal */
exports.setStrikethrough = setStrikethrough;
const setUnderlined = underlined => ({
  _tag: "SetUnderlined",
  underlined
});
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/** @internal */
exports.setUnderlined = setUnderlined;
const toCode = self => {
  switch (self._tag) {
    case "Reset":
      {
        return 0;
      }
    case "SetBold":
      {
        return self.bold ? 1 : 22;
      }
    case "SetColor":
      {
        switch (self.layer) {
          case "foreground":
            {
              return self.vivid ? 90 + color.toCode(self.color) : 30 + color.toCode(self.color);
            }
          case "background":
            {
              return self.vivid ? 100 + color.toCode(self.color) : 40 + color.toCode(self.color);
            }
        }
      }
    case "SetItalicized":
      {
        return self.italicized ? 3 : 23;
      }
    case "SetStrikethrough":
      {
        return self.strikethrough ? 9 : 29;
      }
    case "SetUnderlined":
      {
        return self.underlined ? 4 : 24;
      }
  }
};
/** @internal */
exports.toCode = toCode;
const toEscapeSequence = sgrs => csi("m", sgrs);
exports.toEscapeSequence = toEscapeSequence;
const csi = (controlFunction, sgrs) => {
  const params = Array.from(sgrs).map(sgr => `${toCode(sgr)}`).join(";");
  return `\u001b[${params}${controlFunction}`;
};
