// @effect/printer-ansi/dist/esm/Color.js
/**
 * @since 1.0.0
 */
import * as internal from "./internal/color.js";
// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category constructors
 */
export const black = internal.black;
/**
 * @since 1.0.0
 * @category constructors
 */
export const red = internal.red;
/**
 * @since 1.0.0
 * @category constructors
 */
export const green = internal.green;
/**
 * @since 1.0.0
 * @category constructors
 */
export const yellow = internal.yellow;
/**
 * @since 1.0.0
 * @category constructors
 */
export const blue = internal.blue;
/**
 * @since 1.0.0
 * @category constructors
 */
export const magenta = internal.magenta;
/**
 * @since 1.0.0
 * @category constructors
 */
export const cyan = internal.cyan;
/**
 * @since 1.0.0
 * @category constructors
 */
export const white = internal.white;
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category destructors
 */
export const toCode = internal.toCode;
