// @effect/printer/dist/cjs/Flatten.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.neverFlat = exports.map = exports.isNeverFlat = exports.isFlattened = exports.isFlatten = exports.isAlreadyFlat = exports.flattened = exports.alreadyFlat = exports.FlattenTypeId = void 0;
var internal = _interopRequireWildcard(require("./internal/flatten.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category symbol
 */
const FlattenTypeId = exports.FlattenTypeId = internal.FlattenTypeId;
// -----------------------------------------------------------------------------
// Refinements
// -----------------------------------------------------------------------------
/**
 * Returns `true` if the specified value is a `Flatten`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isFlatten = exports.isFlatten = internal.isFlatten;
/**
 * Returns `true` if the specified `Flatten` is a `Flattened`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isFlattened = exports.isFlattened = internal.isFlattened;
/**
 * Returns `true` if the specified `Flatten` is an `AlreadyFlat`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isAlreadyFlat = exports.isAlreadyFlat = internal.isAlreadyFlat;
/**
 * Returns `true` if the specified `Flatten` is a `NeverFlat`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isNeverFlat = exports.isNeverFlat = internal.isNeverFlat;
// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category constructors
 */
const flattened = exports.flattened = internal.flattened;
/**
 * @since 1.0.0
 * @category constructors
 */
const alreadyFlat = exports.alreadyFlat = internal.alreadyFlat;
/**
 * @since 1.0.0
 * @category constructors
 */
const neverFlat = exports.neverFlat = internal.neverFlat;
// -----------------------------------------------------------------------------
// Combinators
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category mapping
 */
const map = exports.map = internal.map;
