// @effect/printer/dist/cjs/PageWidth.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unbounded = exports.remainingWidth = exports.isUnbounded = exports.isPageWidth = exports.isAvailablePerLine = exports.defaultPageWidth = exports.availablePerLine = exports.PageWidthTypeId = void 0;
var internal = _interopRequireWildcard(require("./internal/pageWidth.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */

// -----------------------------------------------------------------------------
// Models
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category symbol
 */
const PageWidthTypeId = exports.PageWidthTypeId = internal.PageWidthTypeId;
// -----------------------------------------------------------------------------
// Refinements
// -----------------------------------------------------------------------------
/**
 * Returns `true` if the specified value is a `PageWidth`, `false` otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isPageWidth = exports.isPageWidth = internal.isPageWidth;
/**
 * Returns `true` if the specified `PageWidth` is an `AvailablePerLine`, `false`
 * otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isAvailablePerLine = exports.isAvailablePerLine = internal.isAvailablePerLine;
/**
 * Returns `true` if the specified `PageWidth` is an `Unbounded`, `false`
 * otherwise.
 *
 * @since 1.0.0
 * @category refinements
 */
const isUnbounded = exports.isUnbounded = internal.isUnbounded;
// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category constructors
 */
const availablePerLine = exports.availablePerLine = internal.availablePerLine;
/**
 * @since 1.0.0
 * @category constructors
 */
const unbounded = exports.unbounded = internal.unbounded;
/**
 * @since 1.0.0
 * @category constructors
 */
const defaultPageWidth = exports.defaultPageWidth = internal.defaultPageWidth;
// -----------------------------------------------------------------------------
// Utilities
// -----------------------------------------------------------------------------
/**
 * Calculates the remaining width on the current line.
 *
 * @since 1.0.0
 * @category utilities
 */
const remainingWidth = exports.remainingWidth = internal.remainingWidth;
