// @effect/printer-ansi/dist/esm/internal/ansiDoc.js
import * as Doc from "@effect/printer/Doc";
import * as InternalAnsi from "./ansi.js";
/** @internal */
export const beep = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.beep);
/** @internal */
export const cursorTo = (column, row) => Doc.annotate(Doc.empty, InternalAnsi.cursorTo(column, row));
/** @internal */
export const cursorMove = (column, row) => Doc.annotate(Doc.empty, InternalAnsi.cursorMove(column, row));
/** @internal */
export const cursorUp = (lines = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorUp(lines));
/** @internal */
export const cursorDown = (lines = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorDown(lines));
/** @internal */
export const cursorForward = (columns = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorForward(columns));
/** @internal */
export const cursorBackward = (columns = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorBackward(columns));
/** @internal */
export const cursorLeft = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorLeft);
/** @internal */
export const cursorSavePosition = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorSavePosition);
/** @internal */
export const cursorRestorePosition = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorRestorePosition);
/** @internal */
export const cursorNextLine = (rows = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorNextLine(rows));
/** @internal */
export const cursorPrevLine = (rows = 1) => Doc.annotate(Doc.empty, InternalAnsi.cursorPrevLine(rows));
/** @internal */
export const cursorHide = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorHide);
/** @internal */
export const cursorShow = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.cursorShow);
/** @internal */
export const eraseLines = rows => Doc.annotate(Doc.empty, InternalAnsi.eraseLines(rows));
/** @internal */
export const eraseEndLine = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseEndLine);
/** @internal */
export const eraseStartLine = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseStartLine);
/** @internal */
export const eraseLine = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseLine);
/** @internal */
export const eraseDown = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseDown);
/** @internal */
export const eraseUp = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseUp);
/** @internal */
export const eraseScreen = /*#__PURE__*/Doc.annotate(Doc.empty, InternalAnsi.eraseScreen);
