// @effect/printer-ansi/dist/esm/internal/color.js
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/** @internal */
export const black = {
  _tag: "Black"
};
/** @internal */
export const red = {
  _tag: "Red"
};
/** @internal */
export const green = {
  _tag: "Green"
};
/** @internal */
export const yellow = {
  _tag: "Yellow"
};
/** @internal */
export const blue = {
  _tag: "Blue"
};
/** @internal */
export const magenta = {
  _tag: "Magenta"
};
/** @internal */
export const cyan = {
  _tag: "Cyan"
};
/** @internal */
export const white = {
  _tag: "White"
};
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/** @internal */
export const toCode = color => {
  switch (color._tag) {
    case "Black":
      {
        return 0;
      }
    case "Red":
      {
        return 1;
      }
    case "Green":
      {
        return 2;
      }
    case "Yellow":
      {
        return 3;
      }
    case "Blue":
      {
        return 4;
      }
    case "Magenta":
      {
        return 5;
      }
    case "Cyan":
      {
        return 6;
      }
    case "White":
      {
        return 7;
      }
  }
};
