// @effect/printer-ansi/dist/esm/internal/sgr.js
import * as color from "./color.js";
// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/** @internal */
export const reset = {
  _tag: "Reset"
};
/** @internal */
export const setBold = bold => ({
  _tag: "SetBold",
  bold
});
/** @internal */
export const setColor = (color, vivid, layer) => ({
  _tag: "SetColor",
  color,
  vivid,
  layer
});
/** @internal */
export const setItalicized = italicized => ({
  _tag: "SetItalicized",
  italicized
});
/** @internal */
export const setStrikethrough = strikethrough => ({
  _tag: "SetStrikethrough",
  strikethrough
});
/** @internal */
export const setUnderlined = underlined => ({
  _tag: "SetUnderlined",
  underlined
});
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/** @internal */
export const toCode = self => {
  switch (self._tag) {
    case "Reset":
      {
        return 0;
      }
    case "SetBold":
      {
        return self.bold ? 1 : 22;
      }
    case "SetColor":
      {
        switch (self.layer) {
          case "foreground":
            {
              return self.vivid ? 90 + color.toCode(self.color) : 30 + color.toCode(self.color);
            }
          case "background":
            {
              return self.vivid ? 100 + color.toCode(self.color) : 40 + color.toCode(self.color);
            }
        }
      }
    case "SetItalicized":
      {
        return self.italicized ? 3 : 23;
      }
    case "SetStrikethrough":
      {
        return self.strikethrough ? 9 : 29;
      }
    case "SetUnderlined":
      {
        return self.underlined ? 4 : 24;
      }
  }
};
/** @internal */
export const toEscapeSequence = sgrs => csi("m", sgrs);
const csi = (controlFunction, sgrs) => {
  const params = Array.from(sgrs).map(sgr => `${toCode(sgr)}`).join(";");
  return `\u001b[${params}${controlFunction}`;
};
