// @effect/cli/dist/esm/HelpDoc/Span.js
import * as InternalSpan from "../internal/helpDoc/span.js";
/**
 * @since 1.0.0
 * @category refinements
 */
export const isSequence = InternalSpan.isSequence;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isStrong = InternalSpan.isStrong;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isText = InternalSpan.isText;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isUri = InternalSpan.isUri;
/**
 * @since 1.0.0
 * @category refinements
 */
export const isWeak = InternalSpan.isWeak;
/**
 * @since 1.0.0
 * @category constructors
 */
export const empty = InternalSpan.empty;
/**
 * @since 1.0.0
 * @category constructors
 */
export const space = InternalSpan.space;
/**
 * @since 1.0.0
 * @category constructors
 */
export const text = InternalSpan.text;
/**
 * @since 1.0.0
 * @category constructors
 */
export const code = InternalSpan.code;
/**
 * @since 1.0.0
 * @category constructors
 */
export const error = InternalSpan.error;
/**
 * @since 1.0.0
 * @category constructors
 */
export const weak = InternalSpan.weak;
/**
 * @since 1.0.0
 * @category constructors
 */
export const strong = InternalSpan.strong;
/**
 * @since 1.0.0
 * @category constructors
 */
export const uri = InternalSpan.uri;
/**
 * @since 1.0.0
 * @category combinators
 */
export const concat = InternalSpan.concat;
/**
 * @since 1.0.0
 * @category combinators
 */
export const spans = InternalSpan.spans;
