// @effect/platform-bun/dist/cjs/internal/workerRunner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.layer = void 0;
var _WorkerError = require("@effect/platform/WorkerError");
var Runner = _interopRequireWildcard(require("@effect/platform/WorkerRunner"));
var Context = _interopRequireWildcard(require("effect/Context"));
var Deferred = _interopRequireWildcard(require("effect/Deferred"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var Exit = _interopRequireWildcard(require("effect/Exit"));
var FiberId = _interopRequireWildcard(require("effect/FiberId"));
var FiberSet = _interopRequireWildcard(require("effect/FiberSet"));
var Layer = _interopRequireWildcard(require("effect/Layer"));
var Runtime = _interopRequireWildcard(require("effect/Runtime"));
var Scope = _interopRequireWildcard(require("effect/Scope"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const platformRunnerImpl = /*#__PURE__*/Runner.PlatformRunner.of({
  [Runner.PlatformRunnerTypeId]: Runner.PlatformRunnerTypeId,
  start() {
    return Effect.gen(function* () {
      if (!("postMessage" in self)) {
        return yield* new _WorkerError.WorkerError({
          reason: "spawn",
          cause: new Error("not in a Worker context")
        });
      }
      const port = self;
      const run = handler => Effect.uninterruptibleMask(restore => Effect.gen(function* () {
        const scope = yield* Effect.scope;
        const runtime = (yield* Effect.runtime()).pipe(Runtime.updateContext(Context.omit(Scope.Scope)));
        const fiberSet = yield* FiberSet.make();
        const runFork = Runtime.runFork(runtime);
        function onMessage(event) {
          const message = event.data;
          if (message[0] === 0) {
            FiberSet.unsafeAdd(fiberSet, runFork(restore(handler(0, message[1]))));
          } else {
            port.close();
            Deferred.unsafeDone(fiberSet.deferred, Exit.interrupt(FiberId.none));
          }
        }
        function onMessageError(error) {
          Deferred.unsafeDone(fiberSet.deferred, new _WorkerError.WorkerError({
            reason: "decode",
            cause: error.data
          }));
        }
        function onError(error) {
          Deferred.unsafeDone(fiberSet.deferred, new _WorkerError.WorkerError({
            reason: "unknown",
            cause: error.data
          }));
        }
        port.addEventListener("message", onMessage);
        port.addEventListener("messageerror", onMessageError);
        port.postMessage([0]);
        yield* Scope.addFinalizer(scope, Effect.sync(() => {
          port.removeEventListener("message", onMessage);
          port.removeEventListener("messageerror", onError);
        }));
        return yield* restore(FiberSet.join(fiberSet));
      }).pipe(Effect.scoped));
      const send = (_portId, message, transfer) => Effect.sync(() => port.postMessage([1, message], {
        transfer: transfer
      }));
      return {
        run,
        send
      };
    });
  }
});
/** @internal */
const layer = exports.layer = /*#__PURE__*/Layer.succeed(Runner.PlatformRunner, platformRunnerImpl);
