// @effect/printer/dist/esm/internal/docTreeToken.js
const DocTreeTokenTypeId = /*#__PURE__*/Symbol.for("@effect/printer/DocTree/DocTreeToken");
const proto = {
  [DocTreeTokenTypeId]: {
    _A: _ => _
  }
};
/** @internal */
export const empty = /*#__PURE__*/(() => {
  const op = /*#__PURE__*/Object.create(proto);
  op._tag = "EmptyToken";
  return op;
})();
/** @internal */
export const char = char => {
  const op = Object.create(proto);
  op._tag = "CharToken";
  op.char = char;
  return op;
};
/** @internal */
export const text = text => {
  const op = Object.create(proto);
  op._tag = "TextToken";
  op.text = text;
  return op;
};
/** @internal */
export const line = indentation => {
  const op = Object.create(proto);
  op._tag = "LineToken";
  op.indentation = indentation;
  return op;
};
/** @internal */
export const pushAnnotation = annotation => {
  const op = Object.create(proto);
  op._tag = "PushAnnotationToken";
  op.annotation = annotation;
  return op;
};
/** @internal */
export const popAnnotation = /*#__PURE__*/(() => {
  const op = /*#__PURE__*/Object.create(proto);
  op._tag = "PopAnnotationToken";
  return op;
})();
