// @effect/cli/dist/cjs/internal/configFile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeProvider = exports.layer = exports.ConfigFileError = exports.ConfigErrorTypeId = void 0;
var FileSystem = _interopRequireWildcard(require("@effect/platform/FileSystem"));
var Path = _interopRequireWildcard(require("@effect/platform/Path"));
var Cause = _interopRequireWildcard(require("effect/Cause"));
var ConfigProvider = _interopRequireWildcard(require("effect/ConfigProvider"));
var Context = _interopRequireWildcard(require("effect/Context"));
var DefaultServices = _interopRequireWildcard(require("effect/DefaultServices"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var _Function = require("effect/Function");
var Layer = _interopRequireWildcard(require("effect/Layer"));
var InternalFiles = _interopRequireWildcard(require("./files.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const fileExtensions = {
  json: ["json"],
  yaml: ["yaml", "yml"],
  ini: ["ini"],
  toml: ["toml", "tml"]
};
const allFileExtensions = /*#__PURE__*/Object.values(fileExtensions).flat();
/** @internal */
const makeProvider = (fileName, options) => Effect.gen(function* (_) {
  const path = yield* _(Path.Path);
  const fs = yield* _(FileSystem.FileSystem);
  const searchPaths = options?.searchPaths && options.searchPaths.length ? options.searchPaths : ["."];
  const extensions = options?.formats && options.formats.length ? options.formats.flatMap(_ => fileExtensions[_]) : allFileExtensions;
  const filePaths = yield* _(Effect.filter(searchPaths.flatMap(searchPath => extensions.map(ext => path.join(searchPath, `${fileName}.${ext}`))), path => Effect.orElseSucceed(fs.exists(path), () => false)));
  const providers = yield* _(Effect.forEach(filePaths, path => (0, _Function.pipe)(fs.readFileString(path), Effect.mapError(_ => ConfigFileError(`Could not read file (${path})`)), Effect.flatMap(content => Effect.mapError(InternalFiles.parse(path, content), message => ConfigFileError(message))), Effect.map(data => ConfigProvider.fromJson(data)))));
  if (providers.length === 0) {
    return ConfigProvider.fromMap(new Map());
  }
  return providers.reduce((acc, provider) => ConfigProvider.orElse(acc, () => provider));
});
/** @internal */
exports.makeProvider = makeProvider;
const layer = (fileName, options) => (0, _Function.pipe)(makeProvider(fileName, options), Effect.map(provider => Layer.fiberRefLocallyScopedWith(DefaultServices.currentServices, services => {
  const current = Context.get(services, ConfigProvider.ConfigProvider);
  return Context.add(services, ConfigProvider.ConfigProvider, ConfigProvider.orElse(current, () => provider));
})), Layer.unwrapEffect);
/** @internal */
exports.layer = layer;
const ConfigErrorTypeId = exports.ConfigErrorTypeId = /*#__PURE__*/Symbol.for("@effect/cli/ConfigFile/ConfigFileError");
const ConfigFileErrorProto = /*#__PURE__*/Object.assign( /*#__PURE__*/Object.create(Cause.YieldableError.prototype), {
  [ConfigErrorTypeId]: ConfigErrorTypeId
});
/** @internal */
const ConfigFileError = message => {
  const self = Object.create(ConfigFileErrorProto);
  self._tag = "ConfigFileError";
  self.message = message;
  return self;
};
exports.ConfigFileError = ConfigFileError;
