// @effect/cli/dist/cjs/internal/files.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readString = exports.read = exports.parse = exports.fileParsers = void 0;
var FileSystem = _interopRequireWildcard(require("@effect/platform/FileSystem"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var Ini = _interopRequireWildcard(require("ini"));
var Toml = _interopRequireWildcard(require("toml"));
var Yaml = _interopRequireWildcard(require("yaml"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/** @internal */
const fileParsers = exports.fileParsers = {
  json: content => JSON.parse(content),
  yaml: content => Yaml.parse(content),
  yml: content => Yaml.parse(content),
  ini: content => Ini.parse(content),
  toml: content => Toml.parse(content),
  tml: content => Toml.parse(content)
};
/** @internal */
const read = path => Effect.flatMap(FileSystem.FileSystem, fs => Effect.matchEffect(fs.readFile(path), {
  onFailure: error => Effect.fail(`Could not read file (${path}): ${error}`),
  onSuccess: content => Effect.succeed([path, content])
}));
/** @internal */
exports.read = read;
const readString = path => Effect.flatMap(FileSystem.FileSystem, fs => Effect.matchEffect(fs.readFileString(path), {
  onFailure: error => Effect.fail(`Could not read file (${path}): ${error}`),
  onSuccess: content => Effect.succeed([path, content])
}));
/** @internal */
exports.readString = readString;
const parse = (path, content, format) => {
  const parser = fileParsers[format ?? path.split(".").pop()];
  if (parser === undefined) {
    return Effect.fail(`Unsupported file format: ${format}`);
  }
  return Effect.try({
    try: () => parser(content),
    catch: e => `Could not parse ${format} file (${path}): ${e}`
  });
};
exports.parse = parse;
