// @effect/printer-ansi/dist/cjs/Color.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.yellow = exports.white = exports.toCode = exports.red = exports.magenta = exports.green = exports.cyan = exports.blue = exports.black = void 0;
var internal = _interopRequireWildcard(require("./internal/color.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */

// -----------------------------------------------------------------------------
// Constructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category constructors
 */
const black = exports.black = internal.black;
/**
 * @since 1.0.0
 * @category constructors
 */
const red = exports.red = internal.red;
/**
 * @since 1.0.0
 * @category constructors
 */
const green = exports.green = internal.green;
/**
 * @since 1.0.0
 * @category constructors
 */
const yellow = exports.yellow = internal.yellow;
/**
 * @since 1.0.0
 * @category constructors
 */
const blue = exports.blue = internal.blue;
/**
 * @since 1.0.0
 * @category constructors
 */
const magenta = exports.magenta = internal.magenta;
/**
 * @since 1.0.0
 * @category constructors
 */
const cyan = exports.cyan = internal.cyan;
/**
 * @since 1.0.0
 * @category constructors
 */
const white = exports.white = internal.white;
// -----------------------------------------------------------------------------
// Destructors
// -----------------------------------------------------------------------------
/**
 * @since 1.0.0
 * @category destructors
 */
const toCode = exports.toCode = internal.toCode;
