// @effect/platform-bun/dist/esm/internal/workerRunner.js
import { WorkerError } from "@effect/platform/WorkerError";
import * as Runner from "@effect/platform/WorkerRunner";
import * as Context from "effect/Context";
import * as Deferred from "effect/Deferred";
import * as Effect from "effect/Effect";
import * as Exit from "effect/Exit";
import * as FiberId from "effect/FiberId";
import * as FiberSet from "effect/FiberSet";
import * as Layer from "effect/Layer";
import * as Runtime from "effect/Runtime";
import * as Scope from "effect/Scope";
const platformRunnerImpl = /*#__PURE__*/Runner.PlatformRunner.of({
  [Runner.PlatformRunnerTypeId]: Runner.PlatformRunnerTypeId,
  start() {
    return Effect.gen(function* () {
      if (!("postMessage" in self)) {
        return yield* new WorkerError({
          reason: "spawn",
          cause: new Error("not in a Worker context")
        });
      }
      const port = self;
      const run = handler => Effect.uninterruptibleMask(restore => Effect.gen(function* () {
        const scope = yield* Effect.scope;
        const runtime = (yield* Effect.runtime()).pipe(Runtime.updateContext(Context.omit(Scope.Scope)));
        const fiberSet = yield* FiberSet.make();
        const runFork = Runtime.runFork(runtime);
        function onMessage(event) {
          const message = event.data;
          if (message[0] === 0) {
            FiberSet.unsafeAdd(fiberSet, runFork(restore(handler(0, message[1]))));
          } else {
            port.close();
            Deferred.unsafeDone(fiberSet.deferred, Exit.interrupt(FiberId.none));
          }
        }
        function onMessageError(error) {
          Deferred.unsafeDone(fiberSet.deferred, new WorkerError({
            reason: "decode",
            cause: error.data
          }));
        }
        function onError(error) {
          Deferred.unsafeDone(fiberSet.deferred, new WorkerError({
            reason: "unknown",
            cause: error.data
          }));
        }
        port.addEventListener("message", onMessage);
        port.addEventListener("messageerror", onMessageError);
        port.postMessage([0]);
        yield* Scope.addFinalizer(scope, Effect.sync(() => {
          port.removeEventListener("message", onMessage);
          port.removeEventListener("messageerror", onError);
        }));
        return yield* restore(FiberSet.join(fiberSet));
      }).pipe(Effect.scoped));
      const send = (_portId, message, transfer) => Effect.sync(() => port.postMessage([1, message], {
        transfer: transfer
      }));
      return {
        run,
        send
      };
    });
  }
});
/** @internal */
export const layer = /*#__PURE__*/Layer.succeed(Runner.PlatformRunner, platformRunnerImpl);
