// @effect/cli/dist/cjs/BuiltInOptions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.showWizard = exports.showVersion = exports.showHelp = exports.showCompletions = exports.isShowWizard = exports.isShowVersion = exports.isShowHelp = exports.isShowCompletions = exports.builtInOptions = void 0;
var InternalBuiltInOptions = _interopRequireWildcard(require("./internal/builtInOptions.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category options
 */
const builtInOptions = exports.builtInOptions = InternalBuiltInOptions.builtInOptions;
/**
 * @since 1.0.0
 * @category refinements
 */
const isShowCompletions = exports.isShowCompletions = InternalBuiltInOptions.isShowCompletions;
/**
 * @since 1.0.0
 * @category refinements
 */
const isShowHelp = exports.isShowHelp = InternalBuiltInOptions.isShowHelp;
/**
 * @since 1.0.0
 * @category refinements
 */
const isShowWizard = exports.isShowWizard = InternalBuiltInOptions.isShowWizard;
/**
 * @since 1.0.0
 * @category refinements
 */
const isShowVersion = exports.isShowVersion = InternalBuiltInOptions.isShowVersion;
/**
 * @since 1.0.0
 * @category constructors
 */
const showCompletions = exports.showCompletions = InternalBuiltInOptions.showCompletions;
/**
 * @since 1.0.0
 * @category constructors
 */
const showHelp = exports.showHelp = InternalBuiltInOptions.showHelp;
/**
 * @since 1.0.0
 * @category constructors
 */
const showWizard = exports.showWizard = InternalBuiltInOptions.showWizard;
/**
 * @since 1.0.0
 * @category constructors
 */
const showVersion = exports.showVersion = InternalBuiltInOptions.showVersion;
