// @effect/cli/dist/cjs/CommandDescriptor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.wizard = exports.withSubcommands = exports.withDescription = exports.prompt = exports.parse = exports.mapEffect = exports.map = exports.make = exports.getZshCompletions = exports.getUsage = exports.getSubcommands = exports.getNames = exports.getHelp = exports.getFishCompletions = exports.getBashCompletions = exports.TypeId = void 0;
var Internal = _interopRequireWildcard(require("./internal/commandDescriptor.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 * @category symbols
 */
const TypeId = exports.TypeId = Internal.TypeId;
/**
 * @since 1.0.0
 * @category combinators
 */
const getHelp = exports.getHelp = Internal.getHelp;
/**
 * @since 1.0.0
 * @category combinators
 */
const getBashCompletions = exports.getBashCompletions = Internal.getBashCompletions;
/**
 * @since 1.0.0
 * @category combinators
 */
const getFishCompletions = exports.getFishCompletions = Internal.getFishCompletions;
/**
 * @since 1.0.0
 * @category combinators
 */
const getZshCompletions = exports.getZshCompletions = Internal.getZshCompletions;
/**
 * @since 1.0.0
 * @category combinators
 */
const getNames = exports.getNames = Internal.getNames;
/**
 * @since 1.0.0
 * @category combinators
 */
const getSubcommands = exports.getSubcommands = Internal.getSubcommands;
/**
 * @since 1.0.0
 * @category combinators
 */
const getUsage = exports.getUsage = Internal.getUsage;
/**
 * @since 1.0.0
 * @category combinators
 */
const map = exports.map = Internal.map;
/**
 * @since 1.0.0
 * @category combinators
 */
const mapEffect = exports.mapEffect = Internal.mapEffect;
/**
 * @since 1.0.0
 * @category combinators
 */
const parse = exports.parse = Internal.parse;
/**
 * @since 1.0.0
 * @category constructors
 */
const prompt = exports.prompt = Internal.prompt;
/**
 * @since 1.0.0
 * @category constructors
 */
const make = exports.make = Internal.make;
/**
 * @since 1.0.0
 * @category combinators
 */
const withDescription = exports.withDescription = Internal.withDescription;
/**
 * @since 1.0.0
 * @category combinators
 */
const withSubcommands = exports.withSubcommands = Internal.withSubcommands;
/**
 * @since 1.0.0
 * @category combinators
 */
const wizard = exports.wizard = Internal.wizard;
