// @effect/cli/dist/cjs/Usage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.repeated = exports.optional = exports.named = exports.mixed = exports.getHelp = exports.enumerate = exports.empty = exports.concat = exports.alternation = void 0;
var InternalUsage = _interopRequireWildcard(require("./internal/usage.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 * @category combinators
 */
const alternation = exports.alternation = InternalUsage.alternation;
/**
 * @since 1.0.0
 * @category combinators
 */
const concat = exports.concat = InternalUsage.concat;
/**
 * @since 1.0.0
 * @category constructors
 */
const empty = exports.empty = InternalUsage.empty;
/**
 * @since 1.0.0
 * @category constructors
 */
const enumerate = exports.enumerate = InternalUsage.enumerate;
/**
 * @since 1.0.0
 * @category combinators
 */
const getHelp = exports.getHelp = InternalUsage.getHelp;
/**
 * @since 1.0.0
 * @category constructors
 */
const mixed = exports.mixed = InternalUsage.mixed;
/**
 * @since 1.0.0
 * @category constructors
 */
const named = exports.named = InternalUsage.named;
/**
 * @since 1.0.0
 * @category combinators
 */
const optional = exports.optional = InternalUsage.optional;
/**
 * @since 1.0.0
 * @category combinators
 */
const repeated = exports.repeated = InternalUsage.repeated;
