// @effect/typeclass/dist/cjs/SemiApplicative.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.zipWith = exports.zipRight = exports.zipLeft = exports.lift2 = exports.getSemigroup = exports.ap = void 0;
var _Function = require("effect/Function");
var semigroup = _interopRequireWildcard(require("./Semigroup.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 0.24.0
 */

/**
 * Lift a `Semigroup` into 'F', the inner values are combined using the provided `Semigroup`.
 *
 * @category lifting
 * @since 0.24.0
 */
const getSemigroup = F => S => semigroup.make((self, that) => F.map(F.product(self, that), ([a1, a2]) => S.combine(a1, a2)), (self, collection) => F.map(F.productMany(self, collection), ([head, ...tail]) => S.combineMany(head, tail)));
/**
 * Zips two `F` values together using a provided function, returning a new `F` of the result.
 *
 * @param self - The left-hand side of the zip operation
 * @param that - The right-hand side of the zip operation
 * @param f - The function used to combine the values of the two `Option`s
 *
 * @since 0.24.0
 */
exports.getSemigroup = getSemigroup;
const zipWith = F => (0, _Function.dual)(3, (self, that, f) => F.map(F.product(self, that), ([a, b]) => f(a, b)));
/**
 * @since 0.24.0
 */
exports.zipWith = zipWith;
const ap = F => (0, _Function.dual)(2, (self, that) => zipWith(F)(self, that, (f, a) => f(a)));
/**
 * @since 0.24.0
 */
exports.ap = ap;
const zipLeft = F => (0, _Function.dual)(2, (self, that) => zipWith(F)(self, that, _Function.identity));
/**
 * @since 0.24.0
 */
exports.zipLeft = zipLeft;
const zipRight = F => (0, _Function.dual)(2, (self, that) => zipWith(F)(self, that, _Function.SK));
/**
 * Lifts a binary function into `F`.
 *
 * @param f - The function to lift.
 *
 * @category lifting
 * @since 0.24.0
 */
exports.zipRight = zipRight;
const lift2 = F => f => (0, _Function.dual)(2, (self, that) => zipWith(F)(self, that, f));
exports.lift2 = lift2;
