// @effect/cli/dist/cjs/HelpDoc/Span.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.weak = exports.uri = exports.text = exports.strong = exports.spans = exports.space = exports.isWeak = exports.isUri = exports.isText = exports.isStrong = exports.isSequence = exports.error = exports.empty = exports.concat = exports.code = void 0;
var InternalSpan = _interopRequireWildcard(require("../internal/helpDoc/span.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 * @category refinements
 */
const isSequence = exports.isSequence = InternalSpan.isSequence;
/**
 * @since 1.0.0
 * @category refinements
 */
const isStrong = exports.isStrong = InternalSpan.isStrong;
/**
 * @since 1.0.0
 * @category refinements
 */
const isText = exports.isText = InternalSpan.isText;
/**
 * @since 1.0.0
 * @category refinements
 */
const isUri = exports.isUri = InternalSpan.isUri;
/**
 * @since 1.0.0
 * @category refinements
 */
const isWeak = exports.isWeak = InternalSpan.isWeak;
/**
 * @since 1.0.0
 * @category constructors
 */
const empty = exports.empty = InternalSpan.empty;
/**
 * @since 1.0.0
 * @category constructors
 */
const space = exports.space = InternalSpan.space;
/**
 * @since 1.0.0
 * @category constructors
 */
const text = exports.text = InternalSpan.text;
/**
 * @since 1.0.0
 * @category constructors
 */
const code = exports.code = InternalSpan.code;
/**
 * @since 1.0.0
 * @category constructors
 */
const error = exports.error = InternalSpan.error;
/**
 * @since 1.0.0
 * @category constructors
 */
const weak = exports.weak = InternalSpan.weak;
/**
 * @since 1.0.0
 * @category constructors
 */
const strong = exports.strong = InternalSpan.strong;
/**
 * @since 1.0.0
 * @category constructors
 */
const uri = exports.uri = InternalSpan.uri;
/**
 * @since 1.0.0
 * @category combinators
 */
const concat = exports.concat = InternalSpan.concat;
/**
 * @since 1.0.0
 * @category combinators
 */
const spans = exports.spans = InternalSpan.spans;
